### DISCUSSION TITLE
Minimal hardware configuration

### DISCUSSION THREAD

**Discussion by tortelli2** [missionpinball/missionpinball#55] [Q&A]
good morning and happy new year everyone! I'm new but very fascinated by this project, I ask if it is possible to do some tests with a pc (linux) connected directly to an arduino mega board? Thanks Mario

**Comment by cobra18t** [comment]
I think OPP (Open Pinball Project) is going to be the most barebones option for you to start with. It is built for "Blue Pill" boards with the STM32F103C8T6 processor, not an Arduino. Search that processor number and you will find the boards on Amazon or similar. You also need an ST-LINK programmer. 

Here is an example listing that has two boards and a programmer for less than $20: https://www.amazon.com/Aideepen-Programming-Emulator-Downloader-STM32F103C6T6/dp/B09H3988GP/ref=sr_1_11?crid=7ZVOEJ14U531&keywords=STM32F103&qid=1673133528&sprefix=%2Caps%2C548&sr=8-11

Follow the instructions here to load the firmware, but be sure to load the latest (2.4.0.0): https://openpinballproject.wordpress.com/2020/11/03/11-3-2020-loading-stm32-firmware/

OPP has a concept of 4 "wings" on each processor that are a collection of 8 pins each that can perform different functions. The default wing configuration is all inputs, so if you want something different than that, you have to load it via a python script and USB. Here is another guide that takes you through flashing the firmware (already done via the previous link) and loading the configuration: https://pinballmakers.com/wiki/index.php?title=Beginner%27s_Guide_to_STM32_flashing

Once you have the firmware and configuration squared away, you are set for using it with MPF. Here is the OPP landing page in the MPF Docs for more info: https://docs.missionpinball.org/en/stable/hardware/opp/index.html

**Reply by Stefku** [reply]
Alternative source for the hardware: https://de.aliexpress.com/item/1766455290.html?spm=a2g0o.order_list.order_list_main.73.23175c5fKvXoaF&gatewayAdapt=glo2deu

**Reply by cobra18t** [reply]
There is no USB bootloader on the STM32 boards when you buy them--unlike Arduino--so the programmer is required to load firmware.

**Reply by worldpeace-germany** [reply]
To me the question is, depending on skills, if it is worth to spend time on going that way. Building a pinball machine will cost you nevertheless quite a few EUR/Dollars. I started with the Cobrapin board and I am more than happy that I took that approach. Yes it costs more than $20 but worth every penny. Consider as well the time you spend to get you to a point where you would start with such a board. Just my 5ct.

**Comment by tortelli2** [comment]
How many "Blue Pill" can connect to mpf?
In this week i test my 2 board...thank for replyes 😀

**Reply by cobra18t** [reply]
I think USB is the limitation, so maybe 127 boards? Probably more than you are interested in actually connecting. My CobraPin uses 2 blue pills itself.

**Reply by brian90254** [reply]
@tortelli2 Any progress on this? I'm curious about connecting an Arduino to MPF, and not finding any super simple examples.

**Reply by cobra18t** [reply]
> @tortelli2 Any progress on this? I'm curious about connecting an Arduino to MPF, and not finding any super simple examples.

The supported hardware platforms are listed here: https://docs.missionpinball.org/hardware/index.html

I am not aware of a generic Arduino platform for MPF. The Arduino Pinball Controller uses the LISY protocol and uses a combo of an Arduino and Raspberry Pi Zero. The Blue Pill boards that I mentioned before are used in OPP and thus, CobraPin.
